### ISSUE TITLE
classfileResource not set for superclasses

### ISSUE THREAD

**Issue by cdietrich** [classgraph/classgraph#345]
i use Classgraph for JDK classes.
i have the following problem
if a subclass is scanned first and the parent class after
the superclass entry will have `classfileResource == null`
as of `io.github.classgraph.ClassInfo.getOrCreateClassInfo(String, int, Map<String, ClassInfo>)`

my expectation would be that 
`io.github.classgraph.ClassInfo.addScannedClass(String, int, boolean, Map<String, ClassInfo>, ClasspathElement, Resource)`
corrects this in the else branch, (`classInfo.classfileResource = classfileResource;`)
but it does not

**Comment by cdietrich** [comment]
```
public class Main {
	
	public static void main(String[] args) {
        ScanResult result = new ClassGraph()
        		.verbose()
                .enableClassInfo()
                .enableSystemJarsAndModules()
                .scan();
        for (ClassInfo i : result.getAllClasses()) {
        	if (i.getName().equals("java.util.AbstractQueue")) {
        		System.out.println(i.getName());
        		System.out.println(i.getResource());
        	}
        }
    }

}
```

**Comment by cdietrich** [comment]
[classgraph.log](https://github.com/classgraph/classgraph/files/3169401/classgraph.log)

**Comment by cdietrich** [comment]
the same may hold true for classmodifiers

**Comment by lukehutch** [comment]
@cdietrich Released in 4.8.32. Thank you for the bug report and the pull request containing the fix and the testcase!

**Comment by lukehutch** [comment]
PS I added the fix for the modifier bits, thanks for pointing that out too.

**Comment by cdietrich** [comment]
@lukehutch  looks like inner classes are handled differently and have the problem too

```
	private static class Base {
		
	}
	
	public static class Sub extends Base {
		 
	}
	
	@Test
	public void issue345b() {
		try (ScanResult scanResult = new ClassGraph()
				.enableClassInfo()
				.overrideClassLoaders(
						new URLClassLoader(new URL[] { 
								Issue345.class.getResource("/io/github/classgraph/issues/issue345/Issue345$Sub.class"),
								Issue345.class.getResource("/io/github/classgraph/issues/issue345/Issue345$Base.class"),
								Issue345.class.getResource("/io/github/classgraph/issues/issue345/Issue345.class"),
						}))
				.scan()) {
			ClassInfo classInfo = scanResult.getClassInfo("io.github.classgraph.issues.issue345.Issue345$Base");
			System.out.println(classInfo.getName());
			System.out.println(classInfo.getResource());
			assertEquals("protected", classInfo.getModifiersStr());
		}
	}
```

**Comment by lukehutch** [comment]
I'm taking a look at it. By the way, I think a `URLClassLoader` cannot take a list of classfile URLs, only a list of package roots jars or dirs?

**Comment by cdietrich** [comment]
yes. i wonder if there is an easy way to achieve the same

**Comment by lukehutch** [comment]
Really you just have to whitelist the package that the things you want scanned is contained in, e.g.

```
.whitelistPackages(Issue345.class.getPackage().getName())
```

or you can specifically whitelist classes to scan:

```
.whitelistClasses(Issue345.class.getName(), Sub.class.getName(), Base.class.getName())
```

And I looked into the issue with inner classes' `Resource` reference not being set. It is because the inner classes are not being scanned, because you did not call `.ignoreClassVisibility()`:

```
2019-05-11T15:33:48.173-0600	ClassGraph	-------------- Scanning directory: /home/luke/Work/classgraph/target/test-classes/io/github/classgraph/issues/issue345 (took 0.000189 sec)
2019-05-11T15:33:48.173-0600	ClassGraph	---------------- Found classfile within subpackage of whitelisted package: io/github/classgraph/issues/issue345/Issue345$Base.class
2019-05-11T15:33:48.271-0600	ClassGraph	------------------ Parsing classfile (took 0.000205 sec)
2019-05-11T15:33:48.271-0600	ClassGraph	-------------------- Skipping classfile: Class is not public, and ignoreClassVisibility() was not called
2019-05-11T15:33:48.173-0600	ClassGraph	---------------- Found classfile within subpackage of whitelisted package: io/github/classgraph/issues/issue345/Issue345$Sub.class
2019-05-11T15:33:48.271-0600	ClassGraph	------------------ Parsing classfile (took 0.000172 sec)
2019-05-11T15:33:48.271-0600	ClassGraph	-------------------- Found class io.github.classgraph.issues.issue345.Issue345$Sub
2019-05-11T15:33:48.271-0600	ClassGraph	---------------------- Superclass: io.github.classgraph.issues.issue345.Issue345$Base
```

**Comment by cdietrich** [comment]
is the whitelist order kept at scanning?

**Comment by lukehutch** [comment]
Oops though, there's an extra `file:` in the URI that's not supposed to be there, due to a change I just made, I'll fix that. And your code exposed that even with `.overrideClassLoaders()` called, other classloaders are still being sourced for classpath URLs, so I'll fix that too.

**Comment by lukehutch** [comment]
@cdietrich I don't understand your question -- the whitelist is checked, and if the whitelist is empty, everything is considered whitelisted, otherwise anything that matches the whitelist is scanned, and nothing else

**Comment by cdietrich** [comment]
for my tests to make sense i want sub to be scanned/added before base

**Comment by lukehutch** [comment]
Firstly you should check `ClassInfo#isExternalClass()`. Any external class is one whose classfile was not scanned (so the `Resource` object cannot be expected to be set).

As far as ordering, I suggest you sort the classes by name, then process them in reverse lexicographic order. That will ensure that inner classes will be processed before outer classes. Actual `ClassInfo#compareTo(ClassInfo)` will give you standard lexicographic sort order based on `ClassInfo#getName()`.

**Comment by cdietrich** [comment]
my concerns are not about using classgraph. its about the implementation / my test
(how can i make sure sub is processed before base and arrayblockingquere before abstractqueue in the test)

**Comment by lukehutch** [comment]
I suggest you create a DAG, and perform a topological sort on the DAG.

You can create any partial ordering you want: subclasses before superclasses, and inner classes before outer classes. Each arrow in the DAG is an "process-after" relationship. You can add as many constraints as you want, as long as you never break the DAG nature of the graph.

Once you topologically sort the DAG, you will have a total ordering, and you can process this in the sorted order. All arrows between nodes in that ordering will point in one direction.

Does that solve your usecase?

**Comment by cdietrich** [comment]
well i guess we are still talking past each other
we both agree the urlclassloader in my test was a bad idea
but how to replace that
the whitelist command allows to restrict it.
but how can issue345 test 
make sure that the classes are processed in the correct
order so that we can be sure the problem is acually fixed and
not just green because of a fortunate order

**Comment by lukehutch** [comment]
Ah, you were talking about the test itself..

Well, ClassGraph scans in parallel, so is nondeterministic anyway.

I double checked all possible ways a `ClassInfo` could be created from scanning an actual `Resource` though, and the fix you provided catches all of those cases.

The thing you were missing was simply `.ignoreClassVisibility()`. And that you can't expect a `ClassInfo` to have an associated `Resource` if the classfile was not scanned (i.e., if the `Resource` was never actually read).

So I think the current version does what is intended, although I'm fixing a couple of other unrelated things that this turned up.

Does that answer your questions?

**Comment by cdietrich** [comment]
Yes

**Comment by lukehutch** [comment]
Just to explain a little further, sometimes a `ClassInfo` is created as a placeholder -- principally when a reference to a class is discovered from another class, but the referenced class has not yet been scanned. That's the purpose of the `addScannedClass()` method -- to take a placeholder and start filling in the details of the class once the class is itself scanned. So you should simply skip classes that have no associated `Resource`, if you only want the scanned classes.

**Comment by cdietrich** [comment]
thx @lukehutch. looks good now for most cases.
for inner classes i still have the problem that the classmodifier of `32`/`0x0020`
basically leads to nothing in ModifierStr

**Comment by lukehutch** [comment]
@cdietrich I now understand better your question about ordering, and reproducibility of the unit test. I improved the test so that I test in both directions: scanning should be extended to superclasses but not subclasses, and outer classes but not inner classes, because scanning is "extended upwards" (superclasses, interfaces, annotations, and outer classes are scanned if they are cited in a class, even if the cited classes are not themselves whitelisted). Testing in the opposite direction (superclass to subclass) should give a null `Resource` if the class in the "downwards" direction is not whitelisted.

**Comment by lukehutch** [comment]
PS @cdietrich I found several other small issues as a result of your bug report, which are fixed in 4.8.33, which I am releasing now. Thanks for the report!

**Comment by lukehutch** [comment]
> `0x0001` is the public, but here does the `0x0020`/`32` come from?
> is that a inner class modifier?

It's a legacy bit:

https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1-200-E.1

`ACC_SUPER` | `0x0020` | Treat superclass methods specially when invoked                                        		by the `invokespecial` instruction.
-- | -- | --

> thx @lukehutch. looks good now for most cases.
> for inner classes i still have the problem that the classmodifier of `32`/`0x0020` for private static (or 33 for public static)
> basically leads to nothing in ModifierStr

I double-checked the modifier bits, and they are being read correctly from the classfile (i.e. `0x20` for the private inner methods, and `0x33` for the `public` inner method).

You'll notice in the table that there is no `private` modifier bit for classes. That's because a toplevel class can't have a `private` modifier -- there would be no way to call the code in it (if you try it, you will get a compiler error).

I believe that `private` modifiers on inner classes don't result in a modifier bit being stored in the classfile either. This modifier governs the visibility of the synthetic methods and accessor fields that are generated by the compiler, in the case of non-`static` inner classes. In the case of static inner classes, I'm guessing there is no security check performed, even with JDK 9+'s new encapsulation model, so that's probably a security hole if static methods access static state.

I can't see any way to determine that an inner class is private though, since it's not in the modifier bits, and I'm pretty sure it's not stored in any of the other class metadata in the classfile either.

**Comment by lukehutch** [comment]
Actually, I stand corrected, there is a separate modifier field for inner classes, but you have to read that field in the outer class, I think, and apply the bits when the inner class is read: 

https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.7.6-300-D.1-D.1

I'll add this in. Thanks for pointing this out!

**Comment by lukehutch** [comment]
@cdietrich Fixed and released in 4.8.34.

**Comment by cdietrich** [comment]
nice. many thanks
